File: neetcode150/15_greedy/0846_hand_of_straights.cpp
/*
 * =====================================================================
 * LeetCode 846 : Hand of Straights                                  (Medium)
 * https://leetcode.com/problems/hand-of-straights/
 * Category   : Greedy
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Can the cards be rearranged into `groupSize`-long runs of
 *   CONSECUTIVE values? Each card used exactly once.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Greedy from the smallest remaining card: it MUST start a run (no
 *   smaller card can precede it). Consume it and the next
 *   groupSize-1 consecutive values; repeat. A multiset makes "smallest
 *   remaining" an O(log n) operation.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Sorted map (TreeMap) — best
 *   1. Count cards. While cards remain: take the smallest value v,
 *      and for k in v..v+groupSize-1 decrement; a missing/empty value
 *      kills the arrangement.
 * APPROACH 2 · Sort + greedy pointer scan
 *   Sort the array; walk runs in-place, skipping already-used cards.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why "smallest first" is forced, not a choice
 *   The smallest remaining card x cannot be the middle or end of a
 *   run (no x-1 exists among remaining cards — it was consumed by an
 *   earlier run or never existed). It must be the FIRST card of a
 *   run, and that run is fully determined: x..x+groupSize-1. So every
 *   greedy choice is forced — no branching, hence correctness is
 *   immediate and the algorithm is optimal by construction.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(n log n) — n inserts + n removals.
 * MEMORY COMPLEXITY: O(n).
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
public:
    bool isNStraightHand(std::vector<int>& hand, int groupSize) {
        if (hand.size() % groupSize) return false;
        std::map<int, int> cnt;
        for (int c : hand) ++cnt[c];
        while (!cnt.empty()) {
            int first = cnt.begin()->first;             // smallest remaining
            for (int k = 0; k < groupSize; ++k) {
                int v = first + k;
                auto it = cnt.find(v);
                if (it == cnt.end()) return false;      // run can't close
                if (--it->second == 0) cnt.erase(it);
            }
        }
        return true;
    }
};

class Solution_2 {
public:
    bool isNStraightHand(std::vector<int>& hand, int groupSize) {
        if (hand.size() % groupSize) return false;
        std::sort(hand.begin(), hand.end());
        std::vector<bool> used(hand.size(), false);
        for (size_t i = 0; i < hand.size(); ++i) {
            if (used[i]) continue;
            int prev = hand[i], placed = 1;
            used[i] = true;
            for (size_t j = i + 1; j < hand.size() && placed < groupSize; ++j)
                if (!used[j] && hand[j] == prev + 1) {  // next consecutive
                    used[j] = true;
                    prev = hand[j];
                    ++placed;
                }
            if (placed < groupSize) return false;
        }
        return true;
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    struct Case { std::vector<int> hand; int g; bool want; };
    const Case cases[] = {
        {{1, 2, 3, 6, 2, 3, 4, 7, 8}, 3, true},
        {{1, 2, 3, 4, 5}, 4, false},
        {{1, 2, 3, 4, 5, 6}, 2, true},
        {{8, 10, 12}, 3, false}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        auto h1 = c.hand, h2 = c.hand;
        bool r1 = Solution_1().isNStraightHand(h1, c.g);
        bool r2 = Solution_2().isNStraightHand(h2, c.g);
        bool ok = r1 == c.want && r2 == c.want;
        allOk = allOk && ok;
        std::cout << "→ A1=" << (r1 ? "T" : "F") << " A2=" << (r2 ? "T" : "F")
                  << " (want " << (c.want ? "T" : "F") << ") "
                  << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}

File: neetcode150/15_greedy/0846_hand_of_straights.py
"""
=====================================================================
LeetCode 846 : Hand of Straights                                  (Medium)
https://leetcode.com/problems/hand-of-straights/
Category   : Greedy
---------------------------------------------------------------------
PROBLEM
    Can the cards be split into groups of `groupSize` consecutive
    values, each card used once?
---------------------------------------------------------------------
INTUITION
    The smallest remaining card must START a run (nothing smaller
    remains) — the choice is forced, which is what makes the greedy
    correct.
---------------------------------------------------------------------
APPROACH 1 · Sorted counter + greedy runs (best)
    Count cards; repeatedly take the smallest value v and consume
    v..v+groupSize-1; any gap fails.
APPROACH 2 · Sort + pointer scan
    Sort, walk runs skipping already-used cards.
---------------------------------------------------------------------
DEEP DIVE · Forced decisions = provably optimal
    Smallest remaining x can't be mid-run (x-1 consumed or absent),
    so it must open a run, and that run is fully determined. No
    branch to choose, no suboptimal alternative — greedy is optimal
    by construction. Counter dict keys are sorted on demand, making
    "smallest remaining" O(log n) amortized.
---------------------------------------------------------------------
TIME COMPLEXITY : O(n log n)
MEMORY COMPLEXITY: O(n)
=====================================================================
"""
from collections import Counter


class Solution_1:
    def isNStraightHand(self, hand, groupSize):
        if len(hand) % groupSize:
            return False
        cnt = Counter(hand)
        for first in sorted(cnt):
            if cnt[first] == 0:
                continue
            need = cnt[first]
            for v in range(first, first + groupSize):
                if cnt[v] < need:
                    return False
                cnt[v] -= need
        return True


class Solution_2:
    def isNStraightHand(self, hand, groupSize):
        if len(hand) % groupSize:
            return False
        hand.sort()
        used = [False] * len(hand)
        for i in range(len(hand)):
            if used[i]:
                continue
            prev, placed = hand[i], 1
            used[i] = True
            j = i + 1
            while j < len(hand) and placed < groupSize:
                if not used[j] and hand[j] == prev + 1:
                    used[j] = True
                    prev = hand[j]
                    placed += 1
                j += 1
            if placed < groupSize:
                return False
        return True


if __name__ == "__main__":
    import time, tracemalloc
    cases = [([1, 2, 3, 6, 2, 3, 4, 7, 8], 3, True),
             ([1, 2, 3, 4, 5], 4, False),
             ([1, 2, 3, 4, 5, 6], 2, True),
             ([8, 10, 12], 3, False)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for hand, g, want in cases:
        r1, r2 = Solution_1().isNStraightHand(hand[:], g), Solution_2().isNStraightHand(hand[:], g)
        ok &= r1 == r2 == want
        print(f"-> A1={r1} A2={r2} (want {want}) {'PASS' if r1 == want and r2 == want else 'FAIL'}")
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt*1e6:.0f} us | mem: {mem:.0f} KB")
